### PR TITLE
Solr Metrics Receiver

### PR DESCRIPTION
**Pull request by StefanKurek**
Adds ability to collect Solr metrics to ops-agent.

[Scope doc](https://docs.google.com/document/d/1GiSjVfshBnTr4-838bLF2EmCY_SUkz9oakk9bbJP9V0/edit#)

### CODE REVIEW

File: apps/solr.go
Comment (qingling128) [inline apps/solr.go:45]: Can we refactor this part a bit to make use of the common util in https://github.com/GoogleCloudPlatform/ops-agent/pull/370?

File: confgenerator/testdata/invalid/linux/metrics-receiver_solr_malformed_url/input.yaml
Comment (qingling128) [inline confgenerator/testdata/invalid/linux/metrics-receiver_solr_malformed_url/input.yaml:17]: It's preferred for the receiver and pipeline IDs to stay the same as the receiver type in the sample configuration. E.g.  "s/{APP}_metrics/{APP}", "s/{APP}_pipeline/{APP}"

File: confgenerator/testdata/valid/linux/metrics-receiver_solr/input.yaml
Comment (qingling128) [inline confgenerator/testdata/valid/linux/metrics-receiver_solr/input.yaml:22]: Remove `collection_interval` here?

For the context: There should at least be one valid unit test that verifies the absolute minimum user configuration. All optional configurations should be absent for this case including "collection_interval".

File: docs/solr.md
Comment (qingling128) [inline docs/solr.md:22]: It's preferred for the receiver and pipeline IDs to stay the same as the receiver type in the sample configuration. E.g.  "s/{APP}_metrics/{APP}", "s/{APP}_pipeline/{APP}"

File: integration_test/third_party_apps_data/agent/ops-agent/linux/enable_solr
Comment (qingling128) [inline integration_test/third_party_apps_data/agent/ops-agent/linux/enable_solr:25]: Remove `collection_interval`?

File: integration_test/third_party_apps_data/agent/ops-agent/linux/supported_applications.txt
Comment (qingling128) [inline integration_test/third_party_apps_data/agent/ops-agent/linux/supported_applications.txt:3]: There are still some pending PRs for `elasticsearch`. Would this trigger a test failure by including `elasticsearch` early?

File: integration_test/third_party_apps_data/applications/solr/centos_rhel/install
Comment (qingling128) [inline integration_test/third_party_apps_data/applications/solr/centos_rhel/install:15]: Is it possible to use regular yum/apt-get/zypper install?

Chances are a bit higher that users will use the default package manager (e.g. yum, apt-get, zyypper) to install a package instead of a pinned version of the application. The install (aka integration_test/third_party_apps_data/applications/{APP}/{DISTRO}/install) script should try reflect the native installation methods as much as possible.

Same for the other `install` scripts.

File: integration_test/third_party_apps_data/agent/ops-agent/linux/supported_applications.txt
Comment (StefanKurek) [reply integration_test/third_party_apps_data/agent/ops-agent/linux/supported_applications.txt:3]: @qingling128 i didn't actually add elasticsearch. i just moved it to an alphabetical spot

File: integration_test/third_party_apps_data/applications/solr/centos_rhel/install
Comment (StefanKurek) [reply integration_test/third_party_apps_data/applications/solr/centos_rhel/install:15]: @qingling128 it doesn't look like this option is available for solr. their official install instructions just show using a tarball

**Comment by StefanKurek** [comment]
@qingling128 thanks for the feedback! can you take another look when you have a moment?

File: integration_test/third_party_apps_data/applications/solr/centos_rhel/install
Comment (quentinmit) [reply integration_test/third_party_apps_data/applications/solr/centos_rhel/install:15]: Debian 9 and 10 appear to have packages for solr: https://packages.debian.org/buster/solr-tomcat
as do Ubuntu focal and bionic: https://packages.ubuntu.com/focal/solr-tomcat

File: integration_test/third_party_apps_data/applications/solr/centos_rhel/install
Comment (quentinmit) [inline integration_test/third_party_apps_data/applications/solr/centos_rhel/install:26]: Can you edit the existing file in place (e.g. with `sed`) instead of copying the default values here? That will ensure that we are compatible with defaults even if the defaults change.

File: integration_test/third_party_apps_data/applications/solr/centos_rhel/post
Comment (quentinmit) [inline integration_test/third_party_apps_data/applications/solr/centos_rhel/post:5]: Is this necessary for the metrics to exist? Can you add a comment describing what this is doing? (maybe `# Create a database called otel so there are some metrics to scrape`?)

File: docs/solr.md
Comment (qingling128) [reply docs/solr.md:22]: Thanks!

File: integration_test/third_party_apps_data/agent/ops-agent/linux/supported_applications.txt
Comment (qingling128) [reply integration_test/third_party_apps_data/agent/ops-agent/linux/supported_applications.txt:3]: Oops. Missed the last line. I see. Resolving this.

**Review by qingling128 — APPROVED** [review]
LGTM modulo the remaining open threads get resolved.

File: integration_test/third_party_apps_data/applications/solr/centos_rhel/post
Comment (StefanKurek) [reply integration_test/third_party_apps_data/applications/solr/centos_rhel/post:5]: @quentinmit it is. i can throw a comment in there.

File: integration_test/third_party_apps_data/applications/solr/centos_rhel/install
Comment (StefanKurek) [reply integration_test/third_party_apps_data/applications/solr/centos_rhel/install:15]: @quentinmit I don't believe those are what we want to use (and conferred with our dev ops team about this) for a base Solr installation. We are following the instructions of what is in the official install doc https://solr.apache.org/guide/7_0/installing-solr.html

File: integration_test/third_party_apps_data/applications/solr/centos_rhel/post
Comment (StefanKurek) [reply integration_test/third_party_apps_data/applications/solr/centos_rhel/post:6]: @quentinmit done

**Comment by StefanKurek** [comment]
@quentinmit thanks for the feedback. I have updated this so please take another look when you have a chance to resolve the remaining conversations. Thanks!

File: integration_test/third_party_apps_data/applications/solr/centos_rhel/post
Comment (quentinmit) [reply integration_test/third_party_apps_data/applications/solr/centos_rhel/post:6]: Actually I unresolved this, this also seems like something that's important enough to put in the Markdown docs too (something like "Note: No metrics will be reported if the Solr instance does not contain any cores.")

File: integration_test/third_party_apps_data/applications/solr/centos_rhel/install
Comment (quentinmit) [reply integration_test/third_party_apps_data/applications/solr/centos_rhel/install:15]: Debian's docs say that `apt install solr-tomcat` will give you a running Solr with zero configuration: https://wiki.debian.org/Solr

I'll leave it up to @qingling128 what we want to do here. I'd love if we could test both installation methods - the most important part of these integration tests is testing distro-specific details, since otherwise we could just do all the tests on a single distro with manually-installed software. So maybe use the packages on the Debian and Ubuntu releases that have them, and the tarball on the other distros, so we cover them both.

File: integration_test/third_party_apps_data/applications/solr/centos_rhel/post
Comment (StefanKurek) [reply integration_test/third_party_apps_data/applications/solr/centos_rhel/post:6]: @quentinmit I can add that. But basically Solr is unusable without any cores added, so I don't think a customer will ever set it up without one

File: integration_test/third_party_apps_data/applications/solr/centos_rhel/install
Comment (StefanKurek) [reply integration_test/third_party_apps_data/applications/solr/centos_rhel/install:15]: @quentinmit so I tried out that package, but the final installation state was very different than when using the tarball. Also, the documentation that you linked to (https://wiki.debian.org/Solr) refers to features like solr.xml which was deprecated in 2014 and unsupported in 2015. I think it's installing a pretty old version of Solr.

File: integration_test/third_party_apps_data/applications/solr/centos_rhel/post
Comment (StefanKurek) [reply integration_test/third_party_apps_data/applications/solr/centos_rhel/post:6]: @quentinmit added a comment in the docs

Comment (qingling128) [reply integration_test/third_party_apps_data/applications/solr/centos_rhel/post:6]: Thanks!

File: integration_test/third_party_apps_data/applications/solr/centos_rhel/install
Comment (qingling128) [reply integration_test/third_party_apps_data/applications/solr/centos_rhel/install:15]: Started a side conversation around `Test Ops Agent Supported Application Versions`. Shared the doc. Feel free to chime in there. I think we can keep discussing this separately (aka not a blocker for this PR).

**Comment by qingling128** [comment]
The open threads seem to have all been resolved.

Waiting for the integration tests for a final check before merging.

**Pull request merged**
